File: djinn/extract.py
import sys
import rich_click as click
from djinn.utils.file_ops import validate_fq_sam, which_linkedread

@click.command(panel = "Other Tools", no_args_is_help = True, context_settings={"allow_interspersed_args" : False}, epilog = "Documentation: https://pdimens.github.io/djinn/extract")
@click.argument('inputs', required=True, type=click.Path(exists=True, readable=True, dir_okay=False, resolve_path=True), callback = validate_fq_sam, nargs=-1)
def extract(inputs):
    '''
    Extract all barcodes in BAM/FASTQ file(s)

    Inputs must be one SAM/BAM file or two FASTQ files (R1 and R2, can be gzipped). Both FASTQ and SAM/BAM
    inputs expect barcodes to follow the standard haplotagging (BX tag), stlfr (@seq_id#barcode), or tellseq
    (@seq_id:barcode) formats.  Writes to stdout.
    '''
    if len(inputs) == 1:
        for i in extract_barcodes_sam(inputs[0]):
            sys.stdout.write(f"{i}\n")
    else:
        from_ = which_linkedread(inputs[0])
        for i in extract_barcodes_fq(from_, inputs[0], inputs[1]):
            sys.stdout.write(f"{i}\n")

File: djinn/cmd/extract.go
// Copyright © 2026 Pavel Dimens | Github: pdimens
package cmd

import (
	"fmt"
	"runtime"

	"djinn/extract"

	"github.com/spf13/cobra"
)

// preCmd represents the preprocess command
var extractXamCmd = &cobra.Command{
	Use:                   "extract [options] file.bam",
	Short:                 "Extract all unique barcodes",
	Example:               "extract -t 4 bombus.bam > bombus.bc",
	Long:                  "Inputs must be one SAM/BAM file. Writes to stdout.",
	DisableFlagsInUseLine: true,
	SilenceUsage:          true,
	Args: func(cmd *cobra.Command, args []string) error {
		if len(args) == 0 {
			fmt.Printf("%s", cmd.UsageString())
			return fmt.Errorf("please provide a SAM/BAM input")
		}
		if err := cobra.ExactArgs(1)(cmd, args); err != nil {
			return err
		}
		if err := filecheck(args[0]); err != nil {
			return err
		}
		if len(args) == 3 {
			if err := filecheck(args[1]); err != nil {
				return err
			}
		}
		return nil
	},
	RunE: func(cmd *cobra.Command, args []string) error {
		threads, err := cmd.Flags().GetInt("threads")
		if err != nil {
			return err
		}
		maxCores := runtime.NumCPU()
		// clamp between 1 and max system threads
		threads = min(maxCores, max(threads, 1))
		runtime.GOMAXPROCS(threads)

		invalid, err := cmd.Flags().GetBool("invalid")
		if err != nil {
			return err
		}

		return extract.Extract(args[0], invalid, threads)
	},
}

var extractFqCmd = &cobra.Command{
	Use:     "extract <-i> file.fq",
	Short:   "Extract all unique barcodes",
	Example: "extract bombus.R1.fq bombus.R2.fq > bombus.bc",
	Long: "Inputs must be any number of FASTQ files (R1 and R2, can be gzipped). " +
		"Inputs expect barcodes to follow the standard (BX tag), stlfr (@seq_id#barcode), or tellseq " +
		"(@seq_id:barcode) formats. Writes to stdout.",
	DisableFlagsInUseLine: true,
	SilenceUsage:          true,
	Args: func(cmd *cobra.Command, args []string) error {
		if len(args) == 0 {
			fmt.Printf("%s", cmd.UsageString())
			return fmt.Errorf("please provide inputs")
		}

		if err := cobra.MinimumNArgs(1)(cmd, args); err != nil {
			return err
		}
		for _, i := range args {
			if err := filecheck(i); err != nil {
				return err
			}
		}
		return nil
	},
	RunE: func(cmd *cobra.Command, args []string) error {
		invalid, err := cmd.Flags().GetBool("invalid")
		if err != nil {
			return err
		}
		return extract.ExtractFQ(args, invalid)
	},
}

func init() {
	samCmd.AddCommand(extractXamCmd)
	fqCmd.AddCommand(extractFqCmd)

	//---Command line arguments-------------
	extractXamCmd.Flags().BoolP("invalid", "i", false, "Include invalid barcodes")
	extractXamCmd.Flags().IntP("threads", "@", 2, "Decompression threads to use")

	extractFqCmd.Flags().BoolP("invalid", "i", false, "Include invalid barcodes")
}
